job-server/src/spark.jobserver/WebApi.scala

```diff
@@ -531,6 +534,36 @@ class WebApi(system: ActorSystem,
   override def timeoutRoute: Route =
     complete(500, errMap("Request timed out. Try using the /jobs/<jobID>, /jobs APIs to get status/results"))
 
+  def determineProxyUser(aConfig: Config,
+                         authInfo: AuthInfo,
+                         contextName: String): (String, Config) = {
+    if (config.getBoolean("shiro.authentication") && config.getBoolean("shiro.use-as-proxy-user")) {
+      //proxy-user-param is ignored and the authenticated user name is used
+      val config = aConfig.withValue(SparkJobUtils.SPARK_PROXY_USER_PARAM,
+        ConfigValueFactory.fromAnyRef(authInfo.toString))
+      (authInfo.toString + "_" + contextName, config)
+    } else {
+      (contextName, aConfig)
+    }
+  }
+
+  def removeProxyUserPrefix(authInfo: AuthInfo, contextNames: Seq[String]): Seq[String] = {
+    if (config.getBoolean("shiro.authentication") && config.getBoolean("shiro.use-as-proxy-user")) {
+      val userNamePrefix = authInfo.toString + "_"
+      contextNames map { cName =>
```

Comment:
yes, it should be better in performance and readability.
You could even use the prefix a a regex and do something like
```
  val UserNamePrefix = (authInfo.toString + "_(.*)").r
      contextNames collect {
         case UserNamePrefix(cName) => cName
      }

```